File: src/jobagent/mobile/account_privacy.py
"""Owner-authenticated privacy requests; never loads an administrator key.

The separate privacy worker performs exports/erasure. A fresh worker heartbeat
is required by SQL before accepting requests, so a dormant queue cannot pretend
to be a working privacy service. Account operations are not paid features.
"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from uuid import UUID

from fastapi import HTTPException
from pydantic import BaseModel, ConfigDict, Field

from .repository import MobileRepository

MAX_EXPORT_BYTES = 64 * 1024 * 1024
UNAVAILABLE = "Account privacy processing is unavailable. Refresh request status or contact support before trying again."
UNCONFIRMED = "Your privacy request was not confirmed. It may already be queued or processing. Refresh request status before retrying; do not assume deletion has stopped."
MESSAGES = {
    "queued": "Request received. Refresh to check progress.",
    "processing": "Your request is being processed. Do not submit it again.",
    "blocked": "Processing needs operator attention. Your request has been retained.",
    "failed": "Processing stopped safely. Contact support with the request ID.",
    "complete": "Request completed.",
}


class EmptyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ErasureRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    confirmation: str = Field(max_length=32)
    email: str = Field(min_length=3, max_length=254)


async def privacy_rpc(repo: MobileRepository, name: str, payload: dict) -> dict:
    try:
        raw = await repo.request(repo.client, "POST", "/rest/v1/rpc/" + name,
                                 json=payload, max_bytes=128 * 1024)
        result = json.loads(raw)
    except HTTPException as exc:
        if exc.status_code in (401, 403, 429):
            raise
        raise HTTPException(503, UNCONFIRMED if name == "mobile_privacy_request" else UNAVAILABLE) from None
    except (ValueError, UnicodeError):
        raise HTTPException(503, UNCONFIRMED if name == "mobile_privacy_request" else UNAVAILABLE) from None
    if not isinstance(result, dict) or result.get("user_id") != repo.user_id:
        raise HTTPException(503, UNCONFIRMED if name == "mobile_privacy_request" else UNAVAILABLE)
    if result.get("error"):
        code = result["error"]
        status, message = {
            "unverified": (403, "Confirm your account email before continuing."),
            "confirmation": (422, "Enter DELETE MY ACCOUNT and your verified account email exactly."),
            "worker_unavailable": (503, "Privacy processing is offline. This request was not queued. Refresh existing requests or contact support."),
            "erasing": (409, "Account deletion is already pending. Contact support for help."),
            "rate_limited": (429, "A recent export already exists. Download it or retry tomorrow."),
            "not_found": (404, "Privacy request not found."),
        }.get(code, (503, UNAVAILABLE))
        raise HTTPException(status, message)
    return result


def public_request(row: dict) -> dict:
    """Explicit projection: storage paths, worker errors and identifiers stay private."""
    try:
        request_id = str(UUID(row["id"]))
        if row["kind"] not in ("export", "erase") or row["state"] not in MESSAGES:
            raise ValueError()
        ready = row["kind"] == "export" and row["state"] == "complete" and bool(row.get("export_path"))
        if ready:
            ready = datetime.fromisoformat(row["expires_at"].replace("Z", "+00:00")) > datetime.now(timezone.utc)
        return {"id": request_id, "kind": row["kind"], "state": row["state"],
                "created_at": row["created_at"], "completed_at": row.get("completed_at"),
                "expires_at": row.get("expires_at"), "download_ready": ready,
                "message": MESSAGES[row["state"]]}
    except (KeyError, ValueError, TypeError):
        raise HTTPException(503, UNAVAILABLE) from None


async def account_status(repo: MobileRepository) -> dict:
    data = await privacy_rpc(repo, "mobile_privacy_status", {})
    rows = data.get("requests")
    if not isinstance(rows, list) or len(rows) > 50:
        raise HTTPException(503, UNAVAILABLE)
    return {"requests": [public_request(row) for row in rows],
            "capability": {"export": True, "erasure": True,
                           "processing_configured": data.get("worker_ready") is True}}


async def request_export(repo: MobileRepository) -> dict:
    data = await privacy_rpc(repo, "mobile_privacy_request", {"p_kind": "export"})
    return public_request(data.get("request", {}))


async def request_erasure(repo: MobileRepository, body: ErasureRequest) -> dict:
    if (body.confirmation != "DELETE MY ACCOUNT" or not repo.verified_email or
            body.email.strip().casefold() != repo.verified_email.casefold()):
        raise HTTPException(422, "Enter DELETE MY ACCOUNT and your verified account email exactly.")
    data = await privacy_rpc(repo, "mobile_privacy_request", {
        "p_kind": "erase", "p_confirmation": body.confirmation, "p_email": body.email.strip(),
    })
    return public_request(data.get("request", {}))


async def download_export(repo: MobileRepository, request_id: UUID) -> bytes:
    data = await privacy_rpc(repo, "mobile_privacy_status", {})
    rows = data.get("requests", [])
    row = next((row for row in rows if row.get("id") == str(request_id)), None)
    if row is None:
        raise HTTPException(404, "Privacy request not found.")
    if not public_request(row)["download_ready"]:
        raise HTTPException(409, "This export is not ready or has expired. Request a new export.")
    # Never follow a URL/path returned by an untrusted record.
    path = f"{repo.user_id}/{request_id}/account.zip"
    if row.get("export_path") != path:
        raise HTTPException(503, UNAVAILABLE)
    return await repo.request(repo.client, "GET", "/storage/v1/object/account-exports/" + path,
                              max_bytes=MAX_EXPORT_BYTES)

File: web/src/beta/accountPrivacy.ts
export type PrivacyRequest = {
  id: string; kind: "export" | "erase";
  state: "queued" | "processing" | "blocked" | "complete" | "failed";
  created_at: string; completed_at?: string | null;
  expires_at?: string | null;
  error_code?: string | null; message?: string | null; download_ready?: boolean;
};
export type AccountPrivacy = {
  requests: PrivacyRequest[];
  capability: { export: boolean; erasure: boolean; processing_configured: boolean };
};
export const ERASURE_CONFIRMATION = "DELETE MY ACCOUNT";
export function erasurePayload(confirmation: string, acknowledged: boolean, verifiedEmail: string | null) {
  if (!verifiedEmail?.trim()) throw new Error("A verified session email is required. Sign in again with your verified account.");
  if (confirmation !== ERASURE_CONFIRMATION || !acknowledged) throw new Error("Type DELETE MY ACCOUNT exactly and acknowledge the queued erasure request.");
  return { confirmation: ERASURE_CONFIRMATION, email: verifiedEmail };
}
export const pendingPrivacyRequest = (request: PrivacyRequest) => ["queued", "processing", "blocked"].includes(request.state);
export const canDownloadExport = (request: PrivacyRequest) => request.kind === "export" && request.state === "complete" && request.download_ready === true;
export function privacyStateGuidance(request: PrivacyRequest): string {
  if (request.state === "queued") return "Request recorded, awaiting processing. This does not mean the work is complete. Use Refresh requests to check.";
  if (request.state === "processing") return "The service reports processing is in progress. Refresh requests later; no automatic polling is running.";
  if (request.state === "blocked") return "Processing is blocked. Follow the service message or contact the service operator with this request ID. Refresh after the issue is resolved; do not submit duplicates.";
  if (request.state === "failed") return "The service reports this request failed. Read the message and contact the service operator with this request ID before deciding whether to submit another request.";
  return request.kind === "export" ? canDownloadExport(request) ? "The service reports this export is ready to download. Store it securely; it contains private account data." : "The service reports completion, but a download is not ready. Refresh or contact the service operator with this request ID."
    : "The service reports this erasure request complete. Check its message for scope and any data retained; this page does not independently verify deletion.";
}
export function checkedPrivacyRequest(value: unknown): PrivacyRequest {
  const item = value as PrivacyRequest | null;
  if (!item || typeof item.id !== "string" || !/^[a-zA-Z0-9-]+$/.test(item.id) || !["export", "erase"].includes(item.kind) || !["queued", "processing", "blocked", "complete", "failed"].includes(item.state) || typeof item.created_at !== "string"
    || (item.download_ready !== undefined && typeof item.download_ready !== "boolean")
    || [item.message, item.error_code, item.completed_at, item.expires_at].some(field => field != null && typeof field !== "string")) throw new Error("The service returned an unrecognized privacy request. Refresh requests before taking another action.");
  return item;
}
export function checkedAccountPrivacy(value: unknown): AccountPrivacy {
  const data = value as AccountPrivacy | null;
  if (!data || !Array.isArray(data.requests) || !data.capability || [data.capability.export, data.capability.erasure, data.capability.processing_configured].some(flag => typeof flag !== "boolean")) throw new Error("Privacy request status is unavailable. Refresh requests; workspace billing does not control this page.");
  return { ...data, requests: data.requests.map(checkedPrivacyRequest) };
}
